Build a dapp guide (#76)
OpenZeppelin/openzeppelin-docs: components/learn/modules/ROOT/pages/on-dapps.adoc

In this guide, we will see how we can use OpenZeppelin libraries such as https://github.com/OpenZeppelin/openzeppelin-network.js[network-js] and the https://github.com/OpenZeppelin/solidity-loader[hot-loader] to build your dapp, as well as how the https://github.com/OpenZeppelin/starter-kit[Starter Kits] bind all pieces of a dapp together:

  * <<network, Connecting to the Network>>
  * <<hot-loader, Hot-loading your Contracts>>
  * <<starter-kit, Kickstarting a Dapp using Starter Kits>>

If you are in a rush, just skip over to <<starter-kit, the last section>>!

[[network]]
== Connecting to the Network

When building a dapp, the first step is to connect to the Ethereum network. This usually involves the following steps:

  . If the user is running Metamask, use the injected web3 provider to access the network, and monitor for network or account changes

>   . If not, fall back to a connection to a public Infura node

Reviewer: ```suggestion
  . If not, fall back to a public Infura node
```

> To avoid having to code this logic on every dapp, we built https://github.com/OpenZeppelin/openzeppelin-network.js[`@openzeppelin/network`] to take care of it. It ships with a `useWeb3` https://reactjs.org/docs/hooks-intro.html[React hook] for getting a `web3Context` based on the logic above. This `web3Context` object provides access to a connected `web3` instance, network information, the user accounts, a function to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1102.md[request account access], and more.

Reviewer: ```suggestion
To avoid having to code this logic on every dapp, we built xref:network.js::index.adoc[*OpenZeppelin Network JS*] to take care of it. It provides a `useWeb3` https://reactjs.org/docs/hooks-intro.html[React hook] for getting a `web3Context` based on the logic above. This `web3Context` object provides access to a connected `web3` instance, network information, the user accounts, a function to https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1102.md[request account access], and more.
```

Reviewer: (i changed the link from the repo to the docsite)

Author: Cool! I wasn't sure what the link was

> The following snippet shows a small React app component that connects to the network using Metamask (falling back to Infura if not found), prompts the user for account access, and shows the user balance.

Reviewer: ```suggestion
The following snippet is a small React app that connects to the network using Metamask (falling back to Infura if not found), prompts the user for account access, and displays the user balance:
```

> The following snippet shows a small React app component that connects to the network using Metamask (falling back to Infura if not found), prompts the user for account access, and shows the user balance.

Reviewer: Can we have this app hosted somewhere?

Author: Hmm I think we can, I'll give it a shot later

> ```js

Reviewer: highlight.js is supposed to support jsx, I'm not sure why this isn't properly rendered. @frangio ideas?

Reviewer 2: We may need to add the XML language for the JSX highlighting. See this line: https://github.com/highlightjs/highlight.js/blob/master/src/languages/javascript.js#L203

Reviewer: That did it, thanks!

  const web3Context = useWeb3(`wss://mainnet.infura.io/ws/v3/${PROJECT_ID}`);
  const { lib: web3, networkId, accounts, providerName } = web3Context;

  // Methods for requesting accounts access
  const requestAuth = (web3Context) => web3Context.requestAuth();
  const requestAccess = useCallback(() => requestAuth(web3Context), []);

  // Querying account balance
  const [balance, setBalance] = useState(0);
  const getBalance = useCallback(async () => {
    setBalance(
      accounts && accounts.length > 0 
      ? (await web3.eth.getBalance(accounts[0]))
      : 'Unknown')
  }, [accounts, web3.eth, web3.utils]);
  useEffect(() => {
    getBalance();
  }, [accounts, getBalance, networkId]);

  // Show all information to the user
  return (
    <div>
      <div>Network: {networkId || 'No connection'}</div>
      <div>Your address: {accounts ? accounts[0] : 'Unknown'}</div>
      
      {accounts && accounts.length ? (
        <div>Accounts access granted</div>
      ) : !!networkId && providerName !== 'infura' ? (
        <button onClick={requestAccess}>Request Access</button>
      ) : (
        <div>No accounts access</div>
      )}
    </div>
  );
}
```

> 

Reviewer: I feel like this is lacking some sort of conclusion, perhaps something to the effect of 'we just made an app that connects to the ethereum network and integrates with metamask without having to worry about these details'.

> NOTE: You can easily xref:on-gsn.adoc[enable meta-transaction support] in your dapp when using the network library by http://example.com[adding a `gsn` option] to the React hook.

Reviewer: ```suggestion
TIP: You can easily xref:on-gsn.adoc[enable meta-transaction support] in your dapp when using Network JS by xref:network.js:api.adoc#webcontextoptions[adding a `gsn` option] to the React hook.
```

> You can follow a step-by-step tutorial on building a dapp from scratch using the network-js library in the http://example.com[Building a network powered dapp] guide.

Reviewer: ```suggestion
To learn more, head to our step-by-step tutorial on xref:network.js::building-a-dapp-from-scratch.adoc[Building a Dapp from Scratch] using the Networks library.
```

[[hot-loader]]
== Hot-loading Contracts

> Now that we have a working connection to the Ethereum network, we can start interacting with the contracts we have coded and xref:deploy-and-interact.adoc[deployed using the CLI]. While we can create a https://web3js.readthedocs.io/en/v1.2.4/web3-eth-contract.html[regular `web3` contract instance] using the ABI and address, we can instead leverage xref:on-upgrades.adoc[OpenZeppelin upgrades] to hot-load our contracts.

Reviewer: ```suggestion
Now that we have a working connection to the Ethereum network, we can start interacting with the contracts we have coded and xref:deploy-and-interact.adoc[deployed using the OpenZeppelin CLI]. 

While we _could_ create a regular `web3` contract instance xref:contract-loader::index.adoc[from ABI and address], we will instead leverage xref:on-upgrades.adoc[contract upgrades] to hot-load our contracts.
```

> This means that, once we have deployed a contract and loaded it in our running React dapp, the app will **automatically pick up any changes to the Solidity code and upgrade the contract** during development. The https://github.com/OpenZeppelin/solidity-loader[hot loader] will watch all your Solidity contracts, and whenever you recompile, it will automatically upgrade the contract on the local development network - keeping the same address and state, and just changing the code to the new version you wrote.

Reviewer: ```suggestion
This means that once we deploy and load a contract in our running React dapp, the app will **automatically pick up any changes to the Solidity code and upgrade the contract** during development. The https://github.com/OpenZeppelin/solidity-loader[*OpenZeppelin Hot Loader*] will watch all your Solidity contracts for changes, and automatically recompile and upgrade the contract on the local development network - keeping the same address and state, and just changing the code to the new version you wrote.
```

(this is assuming the hot loader _does_ recompile by itself, which seems to be the case)

Author: Yep!

> NOTE: The hot-loader is implemented as at https://webpack.js.org/[webpack] plugin. It will not work if you are using a different bundler for your application.

Reviewer: ```suggestion
NOTE: The Hot Loader is a https://webpack.js.org/[webpack] plugin. It will not work if you are using a different bundler for your application.
```

> To set up the hot-loader, you need to install `@openzeppelin/solidity-loader` and `json-loader`, and add the following https://webpack.js.org/configuration/module/#rule[rule] to your webpack configuration:

Reviewer: I'd consider dropping this and focusing on showcasing how great the loader is, instead of explaining how to use it (which will be covered by its own documentation). This is the approach followed by the Network guide above, and it works great.

Author: Disagree. I think there is value in showing that it is _just_ another webpack plugin to set up, and there is no more magic involved.

Reviewer: Ah, that might be me not being used to webpack then.

```js
test: /\.sol$/,
use: [
  { loader: 'json-loader' },
  {
    loader: '@openzeppelin/solidity-loader',
    options: {
      network: 'development',
      disabled: false
    },
  },
],
```

NOTE: If you are using `create-react-app`, you may need to either eject the application to modify the webpack configuration, or use https://github.com/timarney/react-app-rewired[`react-app-rewired`].

> Then, just load the contract Solidity file from your client code, which will go through the hot-loader. Whenever you modify your `Box` contract source, the loader will automatically upgrade the instance to the latest code for you.

Reviewer: ```suggestion
Then, just load the contract Solidity file from your client code, which will go through the Hot Loader. Whenever you modify your contract source, the loader will automatically upgrade the instance to the latest code for you.
```

```js
// Load web3 context using @openzeppelin/network
const { networkId, lib: web3 } = useWeb3Network('http://127.0.0.1:8545');

// Load contract artifact
const Box = require('../../contracts/Box.sol');

// Load deployed instance
const [box, setBox] = useState(undefined);
if (!box && Box && Box.networks && networkId) {
  const deployedNetwork = Box.networks[networkId.toString()];
  if (deployedNetwork) {
    setBox(new web3.eth.Contract(Box.abi, deployedNetwork.address));
  }
}
```

> You can follow a step-by-step tutorial on enabling hot loader in a React web application at https://forum.openzeppelin.com/t/building-an-openzeppelin-dapp-with-solidity-hot-loader/1843/1[the forum].

Reviewer: ```suggestion
For detailed setup instructions, follow our step-by-step tutorial on https://forum.openzeppelin.com/t/building-an-openzeppelin-dapp-with-solidity-hot-loader/1843[Enabling the Hot Loader in a React web application].
```

[[starter-kit]]
== Kickstarting your Dapp

> As you have seen so far, setting up a dapp may involve a certain deal of boilerplate and configuration, such as creating an OpenZeppelin CLI project, initializing a new webpack client app, setting up network configuration, loading your contracts, etc.

Reviewer: ```suggestion
As you have seen so far, setting up a dapp involves a certain deal of boilerplate and configuration, including creating an OpenZeppelin CLI project, initializing a new webpack client app, configuring network access and loading your contracts.
```

> To kickstart this process, we have built the xref:starter-kits[OpenZeppelin Starter Kits]. Starter Kits are preconfigured dapps with several OpenZeppelin libraries, Infura connections, and https://github.com/ConsenSys/rimble-ui[Rimble UI components], ready to start developing right away.

Reviewer: ```suggestion
To kickstart this process, we have built the xref::starter-kits/index.adoc[*OpenZeppelin Starter Kits*]. Starter Kits are preconfigured dapps with several OpenZeppelin libraries, Infura connections, and https://github.com/ConsenSys/rimble-ui[Rimble UI components], ready to start developing right away.
```

> You can start a new project from a starter kit using the `oz unpack` CLI command:

Reviewer: I just noticed `oz unpack` requires the working directory to be empty, which goes against our recommendation to install the CLI locally (i.e. `npx oz unpack` doesn't work).

Author: Good catch. I'm creating an issue to rethink this, and removing the `npx` from this doc.

Author: https://github.com/OpenZeppelin/openzeppelin-sdk/issues/1374

```console
$ npx oz unpack starter
✓ Kit downloaded and unpacked
The kit is ready to use. 

Quick Start
Run your local blockchain:
> ganache-cli --deterministic
Initialize the OpenZeppelin project:
> openzeppelin init app
Go to the client directory:
> cd client
Run the React app:
> npm run start
Continue in your browser!
More at https://github.com/OpenZeppelin/starter-kit/tree/stable
```

> This will unpack a new preconfigured React dapp, with a network connection set up to both metamask and the local node.

Reviewer: ```suggestion
This will unpack a preconfigured React dapp, with a network connection set up to both Metamask and the local node.
```

> image::StarterKit.png[OpenZeppelin Starter Kit,500]

Reviewer: This image might benefit from some context, perhaps briefly explain what the app does?

> 

Reviewer: Should we mention that the idea behind `starter` is for people to modify it and turn it into their app, like CRA does?

(is that the idea?)

> If you want to learn more about how to use the starter kit, `unpack` the `tutorial` instead of the `starter` kit, which will guide you through the process of deploying and interacting with contracts from a dapp. You can also check out the xref:starter-kits:list[list of all starter kits] available to unpack.

Reviewer: ```suggestion
If you want to learn more about using Starter Kits, `unpack` the xref:starter-kits::tutorial.adoc[`tutorial`] kit instead of `starter`: it will guide you through the process of deploying and interacting with contracts from a dapp. You can also check out the xref:starter-kits::list.adoc[list of all starter kits] available to unpack.
```

> You can read more about Starter Kits in their xref:starter-kits[section of this documentation site].

Reviewer: ```suggestion
```

I think we can remove this, given the multiple links.

## Next Steps

> You now have the initial tools to build a decentralized web-based front-end for your smart contracts. Your next steps should be to:

Reviewer: ```suggestion
You have taken the first steps towards building a decentralized web-based front-end for your smart contracts. Your next tasks should be:
```

>   * xref:unit-testing.adoc[Write tests for your application]

Reviewer: ```suggestion
  * xref:unit-testing.adoc[Writing Tests for your Application]
```

>   * xref:public-staging.adoc[Deploy it to a testnet]

Reviewer: ```suggestion
  * xref:public-staging.adoc[Deploying it to a Testnet]
```
